src/estimagic/parameters/nonlinear_constraints.py

```diff
@@ -0,0 +1,468 @@
+        "tol": c.get("tol", CONVERGENCE_ABSOLUTE_CONSTRAINT_TOLERANCE),
+    }
+
+    return internal_constr
+
+
+# ======================================================================================
+# Transform equality constraints to inequality constraints
+# ======================================================================================
+
+
+def equality_as_inequality_constraints(nonlinear_constraints):
+    """Return constraints where equality constraints are converted to inequality."""
+    constraints = [_equality_to_inequality(c) for c in nonlinear_constraints]
+    return constraints
+
+
+def _equality_to_inequality(c):
+    """Transform a single constraint.
+
+    An equality constaint g(x) = 0 can be transformed to two inequality constraints
+    using (g(x), -g(x)) >= 0. Hence, the number of constraints doubles, and the
+    constraint functions itself as well as the derivative need to be updated.
+
+    """
+    if c["type"] == "eq":
+
+        def transform(x, func):
+            return np.concatenate((func(x), -func(x)), axis=0)
+
+        out = {
+            "fun": partial(transform, func=c["fun"]),
+            "jac": partial(transform, func=c["jac"]),
+            "n_constr": 2 * c["n_constr"],
+            "tol": c["tol"],
+            "type": "ineq",
+        }
+    else:
+        out = c
+    return out
+
+
+# ======================================================================================
+# Helper Functions
+# ======================================================================================
+
+
+def _process_selector(c):
+    if "selector" in c:
+        selector = c["selector"]
+    elif "loc" in c:
+
+        def selector(params):
+            return params.loc[c["loc"]]
+
+    elif "query" in c:
+
+        def selector(params):
+            return params.query(c["query"])
+
+    else:
+        selector = _identity
+    return selector
+
+
+def _compose_funcs(f, g):
+    return lambda x: g(f(x))
+
+
+def _identity(x):
+    return x
+
+
+# ======================================================================================
+# Jacobian helper functions
+# ======================================================================================
+
+
+def _extend_jacobian(jac_mat, selection_indices, n_params):
+    """Extend Jacobian on selected parameters to full params.
+
+    Jacobian of constraints is defined on a selection of the parameters, however, we
+    need the Jacobian on the full params. Since the Jacobian is trivially zero at the
+    non-selected params we can simply fill a zero matrix.
+
+    """
+    jac_sparse = np.zeros((jac_mat.shape[0], n_params))
+    jac_sparse[:, selection_indices] = jac_mat
+    return jac_sparse
+
+
+def _get_selection_indices(params, selector):
+    """Get index of selected flat params and number of flat params."""
+    registry = get_registry(extended=True)
+    flat_params = tree_just_flatten(params, registry=registry)
```

Comment:
Use `tree_flatten` to get a `treedef` and then use that `treedef` in `tree_unflatten`. Currently original pytrees can substitute for `treedefs` in pybaum but this will change in the future!